### ISSUE TITLE
Feature requests: Silence detection, time based splitting

### ISSUE THREAD

**Issue by audiamus** [Mbucari/AAXClean#2]
This sounds like a most interesting project.

I am currently using FFmpeg throughout in my project [AAX Audio Converter](https://github.com/audiamus/AaxAudioConverter).

Besides copy/decrypt operations for AAX to MP4 (M4A/M4B), conversion to MP3 is crucial as well. And due to some apparent flaws in the metadata structure for some AAX files, pure AAC stream extraction is also being used. 

However, an essential feature of AAX Audio Converter are the file splitting options. These are not only most important for me and an important reason to create this converter in the first place, but many app users also turned to like them.

Proper file splitting first requires silence detection in the audio stream, governed by parameters level and duration. The output of FFmpeg is somewhat limited here, as it is all text and the field width is too small for long books. But there are ways around that.

From the detected silences I build some sort of cue sheet and then split the input or an intermediate file. If chapters are to play a role in this splitting operation, I adjust the chapter marks to silence positions, particularly necessary with the "named chapters" json info, as those chapter marks are usually inaccurate.      

So, to be able to replace FFmpeg in my project with your lib some time in the future, silence detection and time-based splitting would be needed. It would be great if you can give this a thought. The advantage of an integrated solution versus extra processes is obvious.

**Comment by Mbucari** [comment]
I have an immediate plan to add time based splitting for mp3s. Silence based splitting is doable since the audio filter sees the raw wave stream, but I'll have to give it some thought. How sophisticated does it need to be? Would something as simple as x seconds below y threshold be sufficient?

You're right about the FFMpeg limitations, they are why I started this project. Also, it's slow!

**Comment by Mbucari** [comment]
> And due to some apparent flaws in the metadata structure for some AAX files, pure AAC stream extraction is also being used.

Can you elaborate on the flaw you found? I found a flaw that caused cover not to display on android. The only FFMpeg workaround was to extract the aac stream and mux a fresh mp4. I fixed it in AAXClean's m4b lossless conversion. I documented it [Here](https://superuser.com/questions/1660268/ffmpeg-mp4-cover-art-shows-on-windows-but-not-in-android).

**Comment by audiamus** [comment]
> How sophisticated does it need to be? Would something as simple as x seconds below y threshold be sufficient?

For silence detection with FFmpeg I set a db level and a minimum duration and get all the silences in msec resolution (at least for short enough files), start and duration. That's what I need. I would still do the splitting myself in AAX Audio Converter, particularly with all the heuristic adjustments developed over time. Split parameters are start time and duration.     

> Can you elaborate on the flaw you found? 

There are a few. Some spring to mind immediately:

- Longer books, something like over 15 hours, won't import into iTunes properly (M4B). Extracting the AAC audio stream and rebuilding the MP4 file from scratch seem to solve that.

- Some 44.1kHz books appear to have an ill-formatted meta tag. Brutal patching solves that. https://github.com/KrumpetPirate/AAXtoMP3/issues/116

- Very long books, something like over 35 hours, exhaust FFmpeg 32bit memory address space so I use a recent 64bit build in these cases. But the older 32bit FFmpeg build is much faster for AAX to MP4 decrypt/copy operations. 

-  Nero and Quicktime chapters for MP4. Giving up on FFmpeg here, I now set all meta data, including chapters, for MP4 and MP3 with https://github.com/Zeugma440/atldotnet (Not many players support MP3 chapters but some do.)

- I also found an issue with MP3 track length not recognized by some players. Not really solved. I am not sure whether that exists as a meta tag at all. 

One more thing, well, when you look closer at it again you notice there is more: Users had asked me to add options for variable bit rate or to reduce constant bit rate. This makes every conversion a transcoding, of course. AAX Audio Converter also supports the older AA file format (Audible Format 4, basically MP3), And almost all the options work with both input (AAX=MP4, AA=MP3 )and output (M4A/M4B=MP4 and MP3) formats.

**Comment by Mbucari** [comment]
@audiamus 
Try downloading and using AAXClean. All of the symptoms you're describing sound like they're related to the problem I had and fixed.
EDIT: I read that article you linked and I discovered the same problem. AAXClean sets DependsOnCoreCoder flag to 0. But try it anyway, I think you'll be impressed by the speed. It decrypts at about 100 MBps.

**Comment by Mbucari** [comment]
Well, I was able to implement the silence detection. It works, but unfortunately it's painfully slow. The problem is the aac decoder, faad2. It's just slow, about 1/4 as fast as FFMpeg's decoder. When using it to convert to mp3 it's fine because the mp3 encoder is slower than the aac decoder and they're on different threads. But for this operation it should be fast.

**Comment by audiamus** [comment]
> Try downloading and using AAXClean.

Sure. I will give it a try over the weekend. 

> Well, I was able to implement the silence detection. It works, but unfortunately it's painfully slow.

Thanks a lot nonetheless There is no time pressure feature-wise from my side. I will be busy with other things for several weeks to come.

**Comment by Mbucari** [comment]
@audiamus 
I've added multipart mp3 and multipart mp4 support, although multipart m4b may not yet be stable. I added support for FDK AAC decoder, but Silence detect is still slow (about 1/3 as fast as FFMpeg). I'm going to look at using FFMpeg's avcodec library to make a managed aac decoder. It won't be part of AAClean, but a standalone optional AAC decoder. Unfortunately that would require all FFMpeg DLLs to work since they are statically linked, so the file size would be large. But it would be a single process.

**Comment by Mbucari** [comment]
Multipart mp4 is ready for prime time. At this time AAXClean has your whole wish list, but silence detect is still slow.

**Comment by audiamus** [comment]
> Multipart mp4 is ready for prime time

Excellent. 

I have to admit I didn't find the time to do much testing at all, so I only ran the standard AAX to M4B conversion once. And it did its job as it was supposed to do, of course. Time wise, however, it was significantly slower than my current implementation with FFmpeg. I tested on my old development machine with a 10 hour book in 44.1kHz, 128 kb/s and stopwatched it. AAXClean took about 18 sec while AAX Audio Converter only 12, the latter including all the meta data rewriting and both MP4 chapter variants. 

Maybe the advantage is down to that older FFmpeg 32bit build I'm using. You can try for yourself with the current release of AAX Audio Converter. Switch off all the fancy options and run a bare bones conversion. 

A further observation, checking with MediaInfo: AAXClean left the original proprietary meta tags in place which I think is appropriate. It should be left to application software or the user to customize those. However, one of the two chapter variants was missing altogether. I can't really remember how distinguish the two but I know one is embedded into the audio stream and perhaps that's the one missing.

**Comment by Mbucari** [comment]
>Time wise, however, it was significantly slower than my current implementation with FFmpeg.

Strange. That has not been my experience at all. I'll try out AAX Audio Converter to compare.

> However, one of the two chapter variants was missing altogether.

FFMpeg adds nero chapters, which aren't present in the aax file and are redundant (and can possibly flummox some media players). AAXClean retains the original timed text stream chapters but doesn't add nero chapters. It's equivalent to FFMpeg's -movflags disable_chpl.

EDIT: I just tested AAX Audio Converter with a 1.3 GB book (44kHz, 218 Kbps). It decrypted in 21 seconds. AAXClean did it in 15 seconds.

```C#
var sw = new Stopwatch();
sw.Start();

var aax = new AaxFile(@"C:\Users\mbuca\AppData\Local\Packages\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\LocalState\Content\The Evening and the _0593289641_LC_128_44100_Stereo.aax");
aax.SetDecryptionKey("XXXXXXXX");

var outFile = File.OpenWrite(@"C:\Users\mbuca\Music\Audiobook\Ken Follett\evening and morning.m4b");

aax.ConvertToMp4a(outFile);
sw.Stop();

Console.WriteLine("Encoding Time: {0}", sw.Elapsed.ToString("mm\\:ss"));
```
When splitting by chapters, AAX Audio Converter processed the same file in 45 seconds. AAXClean did it in 20 seconds.

```C#
Stopwatch sw = new Stopwatch();
sw.Start();

var aax = new AaxFile(@"C:\Users\mbuca\AppData\Local\Packages\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\LocalState\Content\The Evening and the _0593289641_LC_128_44100_Stereo.aax");
aax.SetDecryptionKey("XXXXXXXX");

var chapterInfo = aax.GetChapterInfo();
aax.ConvertToMultiMp4a(chapterInfo, NewSplitFile);
sw.Stop();

Console.WriteLine("Encoding Time: {0}", sw.Elapsed.ToString("mm\\:ss"));

void NewSplitFile(NewSplitCallback newSplit)
{
	var dir = @"C:\Users\mbuca\Music\Audiobook\Ken Follett\The Kingsbridge Novels\BookSplit";

	var file = newSplit.Chapter.Title.Replace(":", "") + ".m4b";

	newSplit.OutputFile = File.OpenWrite(Path.Combine(dir, file));
}
```

**Comment by Mbucari** [comment]
The only part that is slower than FFMpeg is silence detection, and it's slower by a wide margin. As I said earlier, I'm going to try writing a c# wrapper for libavcodec which would speed up silence detection a lot. It would still require ffmpeg dlls, but everything would be inside one process.

And in case you didn't know, you can supply a custom ChapterInfo to to the converter. So if you wanted to parse the book's content_metadata_[BOOK_ID].json file for chapter titles and build your own, you can do that.

**Comment by audiamus** [comment]
> I just tested AAX Audio Converter with a 1.3 GB book (44kHz, 218 Kbps). It decrypted in 21 seconds. AAXClean did it in 15 seconds.

Strange indeed.

I did a few more tests with other books, maximum size around 750 MB (I don't have bigger ones because I always download individual parts, so that AAX Audio Converter can parallelize the work, even in single file mode.) But the same pattern for every book: time for AAX Converter vs AAXClean remains 2 : 3.

My setup for AAXClean is exactly like yours. For AAX Audio Converter, all options on the "Chapter" page in Basic Settings should be set to off, no or 0. Otherwise additional conversion phases may be executed to create a better result, but will take additional time. All other settings could be a left at default values I would think, except for "single file" mode and M4A/M4B format.

Anyway, it's not that essential at the moment as I am still occupied with other tasks and that will not change for several weeks to come.

**Comment by Mbucari** [comment]
> all options on the "Chapter" page in Basic Settings should be set to off, no or 0. 


Here are the settings I used for chapters (all other settings are default):
![Converter settings](https://user-images.githubusercontent.com/37587114/127218875-e060f6bd-77cb-4bf8-ba61-62a874793093.png)

Using those settings, I decrypted a single part of a 465 MB, 44kHz, 128 kbps book. Aax Audio Converter took 15 seconds to decrypt it, and AAXClean, using the multipart code above, took 7 seconds.

I tried again with a 293 MB, 22 kHz, 64 kbps book part.  Aax Audio Converter took 11 seconds to decrypt it, and AAXClean took 5 seconds.

I also did a multithreaded test using the below code.  On all 3 parts of the 44kHz book, Aax Audio Converter took 39 seconds and AAXClean took 10 seconds. On all 3 parts of the 22 kHz book, Aax Audio Converter took 24 seconds and AAXClean took 7 seconds.

Multithreaded code:
```C#
Stopwatch sw = new Stopwatch();
sw.Start();

int numThreads = 3;

var bookParts = new string[]
{
	@"C:\Users\mbuca\AppData\Local\Packages\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\LocalState\Content\The Evening and the _B08GMCC7PZ_LC_128_44100_Stereo.aax",
	@"C:\Users\mbuca\AppData\Local\Packages\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\LocalState\Content\The Evening and the _B08GLZNVRB_LC_128_44100_Stereo.aax",
	@"C:\Users\mbuca\AppData\Local\Packages\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\LocalState\Content\The Evening and the _B08GM8VRDB_LC_128_44100_Stereo.aax"
};

var chapterInfos = new ChapterInfo[]
{
	GetChapterInfo("B08GMCC7PZ"),
	GetChapterInfo("B08GLZNVRB"),
	GetChapterInfo("B08GM8VRDB")
};

var decryptionTasks = new List<Task>();

for (int i = 0; i < numThreads; i++)
{
	var aax = new AaxFile(bookParts[i]);
	aax.SetDecryptionKey("xxxxxxxx");

	var chapterInfo = chapterInfos[i];

	decryptionTasks.Add(Task.Run(()=> aax.ConvertToMultiMp4a(chapterInfo, NewSplitFile)));
}

Task.WhenAll(decryptionTasks).GetAwaiter().GetResult();

sw.Stop();

Console.WriteLine("Encoding Time: {0}", sw.Elapsed.ToString("mm\\:ss"));

private static void NewSplitFile(NewSplitCallback newSplit)
{
	var dir = @"C:\Users\mbuca\Music\Audiobook\Ken Follett\The Kingsbridge Novels\BookSplit";

	var file = newSplit.Chapter.Title.Replace(":", "").Replace("?","") + ".m4b";

	newSplit.OutputFile = File.OpenWrite(Path.Combine(dir, file));
}

private static ChapterInfo GetChapterInfo(string bookId)
{
	var file = $"C:\\Users\\mbuca\\AppData\\Local\\Packages\\AudibleInc.AudibleforWindowsPhone_xns73kv1ymhp2\\LocalState\\filescache\\content_metadata_{bookId}.json";

	var json = File.ReadAllText(file);
	var jsonSerializer = JsonSerializer.Deserialize<Root>(json);

	var chapterInfo = new ChapterInfo();

	foreach (var c in jsonSerializer.content_metadata.chapter_info.chapters)
	{
		chapterInfo.AddChapter(c.title, TimeSpan.FromMilliseconds(c.length_ms));
	}

	return chapterInfo;
}
```

**Comment by Mbucari** [comment]
Based on our drastically different experiences, I wonder if environment are different.

I'm running AAXClean on Windows 10 and testing in VS 2019 using a .net 5.0 console app. I'm running the latest Aax Audio Converter release build on the same windows 10 machine.

**Comment by Mbucari** [comment]
Based on our drastically different experiences, I wonder if our environments are different.

I'm running AAXClean on Windows 10 and testing in VS 2019 using a .net 5.0 console app. I'm running the latest Aax Audio Converter release build on the same windows 10 machine.  

One thought I had is maybe your .net environment isn't using your system's AES crypto hardware acceleration, but your FFMpeg somehow is able to use hardware acceleration.

In the same application you are using to test AAXClean, add the following code: My computer took 17 seconds to run that; how long did yours take?

```C#
Stopwatch sw = new Stopwatch();
sw.Start();

byte[] mockAacFrame = new byte[360];

byte[] fakeKey = new byte[16];

for (int i = 0; i < 10000000; i++)
{
	DecryptInPlace(fakeKey, fakeKey, mockAacFrame);
}
sw.Stop();

Console.WriteLine("Encoding Time: {0}", sw.Elapsed.ToString("mm\\:ss"));

static void DecryptInPlace(byte[] key, byte[] iv, byte[] encryptedBlocks)
{
	using var aes = System.Security.Cryptography.Aes.Create();
	aes.Mode = System.Security.Cryptography.CipherMode.CBC;
	aes.Padding = System.Security.Cryptography.PaddingMode.None;

	using var cbcDecryptor = aes.CreateDecryptor(key, iv);

	using var cStream = new System.Security.Cryptography.CryptoStream(new MemoryStream(encryptedBlocks), cbcDecryptor, System.Security.Cryptography.CryptoStreamMode.Read);

	cStream.Read(encryptedBlocks, 0, encryptedBlocks.Length & 0x7ffffff0);
}
```

EDIT: one other thought I had, when you're testing AAXClean, are you using the same input/output storage mediums as when you're running Aax Audio Converter? If you're writing to an SSD with Aax Audio Converter, but writing to an HDD with AAXClean, that could account for the speed difference.

**Comment by audiamus** [comment]
> One thought I had is maybe your .net environment isn't using your system's AES crypto hardware acceleration, but your FFMpeg somehow is able to use hardware acceleration.

Interesting. Might be.

> In the same application you are using to test AAXClean, add the following code: My computer took 17 seconds to run that; how long did yours take?

25 sec.

My machine is a 10 year old i7 (in the upper range at that time), still with Win7. I'm running AAXClean under a .Net 5 console app, built with VS2019. For my test, source and destination are identical for AAX Audio Converter and AAXClean, I feed from an NAS via 1Gbit ethernet and save to a local RAID1 HDD. AAX Audio Converter installed from the GitHub package. Settings on the chapter page like yours, all off.

**Comment by Mbucari** [comment]
I profiled AAXClean and found that I/O uses more cycles than I had expected, and I'm guessing you're CPU-bound when running AAXClean. I split reading/decoding and writing into different threads within ConvertToMultiMp4a. If you truly were CPU-bound, this change should be a marked improvement. It's not the best way to accomplish multi-threading, but it should at least tell me if we found the problem.

When you get a chance, try ConvertToMultiMp4a in the latest version.

**Comment by Mbucari** [comment]
I made a change that dramatically improved the decryption speed. Can you test it out?

**Comment by audiamus** [comment]
I will but can't make promises when. As I said I'm quite busy with other tasks and also going on a much needed vacation next week.

**Comment by Mbucari** [comment]
@audiamus No worries. Get to it when you can. I just wanted to check in and let you know I'm still working on this.

**Comment by audiamus** [comment]
_A few weeks later..._

Just a quick update from my side:

Today I have successfully downloaded my first aaxc file, using some new code of my own. It's based on the incredible work of @omarroth, @mkb79 and @rmcrackan, of course, plus Vesselin Bontchev, who appears to have found the solution in the first place.

Having deciphered key and initialization vector from the license response, I fed it all to AAXClean. It worked instantaneously and was fast as well. Thanks so much!

This is all still just lab code, far from any alpha stage, and I have only vague ideas at the moment how a future direct download option for AAX Audio Converter may look like. It will make use of AAXClean, that's for sure. And as you know, silence detection plays an important role in proper silence based file splitting, one of the strong points of my little tool, and the reason for this thread. So, that has to be incorporated as well. It remains to be seen how the different processing steps will be combined in the end.

**Comment by Mbucari** [comment]
Great! I've been pretty busy with other things and probably won't be able to work on the silence detection for another month.  It does currently work, it's just slow. Right now AAXClean only supports 64-bit.  Is 32-bit important to you? Supporting 32-bit is easy enough, I just need to add 32-bit native dlls.

**Comment by audiamus** [comment]
> Is 32-bit important to you?

I would think so, Currently, AAX Audio Converter can run as a 32bit process and I would like to keep that option. I have an old laptop with 32 bit Windows where I test this variant. But I don't know how many people are actually using the app in a 32 bit environment, apart from one or two who explicitly told me they do.

**Comment by rmcrackan** [comment]
You’re nicer than me. I’ve added a lot of weird edge case support to Libation but I think that’s where I’d draw the line. lol

**Comment by Mbucari** [comment]
OK, I've done pretty much everything we've talked about. Changed:

- All Mp3 conversion and silence detection has been moved to a new project [AAXClean.Codecs](https://github.com/Mbucari/AAXClean.Codecs)
- AAXClean is completely platform independent now, but AAXClean.Codecs is windows-only.
- Both AAXClean and AAXClean.Codecs are ANU CPU
- I built a small aac decoder lib out of ffmpeg and included it in AAXClean.Codecs. All decoding operations are now MUCH faster.
